Added function for finding equation of Ellipse using slope as parameter and faster method for calculation of circumference of ellipse

avishrivastava11 opened the pull request:
Added function for finding equation of Ellipse using slope as parameter.
Added another method `Ellipse_Cirumference` for calculation of circumference of ellipse.
Added a new method called `are_collinear`
Pluralized the following methods
`direction_ratio` -> `direction_ratios`
`direction_cosine` -> `direction_cosines`

Fixes #2815
Fixes #6952
Fixes #7713

This PR uses the approach to finding equation of ellipse using slope, length of semi minor axis and length of semi major axis as inputs given [here](https://math.stackexchange.com/questions/108270/what-is-the-equation-of-an-ellipse-that-is-not-aligned-with-the-axis/646971#646971)
This could be an added functionality to the equation finding method in class `Ellipse`.
Thanks to @smichr  for providing the approach.

Please take a look at this PR and suggest changes. I will be glad to implement them.
Thanks.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* geometry
   * added function `equation_using_slope` for finding equation of Ellipse using slope as parameter
<!-- END RELEASE NOTES -->

gdp1807 committed:
Added a mechanism to check linearity in linear_eq_to_matrix

added a mechanism to check if the equations passed
to linear_eq_to_matrix method are linear or not. If they
are not linear then appropriate error messages are raised.

avishrivastava11 committed:
added function for finding equation of Ellipse using slope as parameter

avishrivastava11 commented:
@smichr  Sir, since you were the one who brought up the idea of adding this functionality, can you take a look?

@jksuom  Sir can you take a look as well?

avishrivastava11 committed:
added faster method to calculate circumference using @smichr 's issue and tips

avishrivastava11 commented:
@smichr @jksuom 
I have added another method `Ellipse_Cirumference` for calculation of circumference of ellipse. This method is much faster as compared to the other method given already. This implementation is done on the lines given by @smichr  [here](https://github.com/sympy/sympy/issues/6952). So this PR can close issue #6952 also ( issue raised by @smichr ).

```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import Ellipse
>>> e1 = Ellipse((0,0), 3, 1)
>>> e1.circumference.n()
13.3648932205553
>>> e1.Ellipse_Circumference()
13.3648932205553
```

@smichr  @jksuom  Can you please take a look?

avishrivastava11 commented:
@certik @jksuom  @smichr  Can you review this PR?

certik commented on sympy/geometry/ellipse.py:
```diff
@@ -433,6 +455,69 @@ def equation(self, x='x', y='y'):
+        https://math.stackexchange.com/questions/108270/what-is-the-equation-of-an-ellipse-hat-is-not-aligned-with-the-axis/646971
+        """
+
+        a, b, x, y, m, M, x_c, y_c, s = var('a,b,x,y,m,M,x_c,y_c,s')
```
don't use `var`, but `symbols`. `var` injects the symbols into the parent scope.

certik replied:
More over, the user should be able to set these symbols, so that he can later reference them. Probably passing them as arguments, optionally?

certik replied:
@avishrivastava11, I agree with what @asmeurer said.

certik commented on sympy/geometry/ellipse.py:
```diff
@@ -323,6 +323,28 @@ def circumference(self):
+    def Ellipse_Circumference(self):
+        """
+        Computes the circumference of an ellipse using length of semi-axes as a and b.
+        Require a >= 0 and b >= 0. This is much faster than the circumference method given above.
```
what do you mean by `above`?

certik commented on sympy/geometry/ellipse.py:
```diff
@@ -323,6 +323,28 @@ def circumference(self):
+        """
+        Computes the circumference of an ellipse using length of semi-axes as a and b.
+        Require a >= 0 and b >= 0. This is much faster than the circumference method given above.
+        Relative accuracy is about 0.5 ** 53.
```
You should mention that it computes the result numerically, using double precision.

certik commented on sympy/geometry/ellipse.py:
```diff
@@ -323,6 +323,28 @@ def circumference(self):
         else:
             return 4 * self.major * elliptic_e(self.eccentricity**2)
 
+    def Ellipse_Circumference(self):
```
use lowercase: `ellipse_circumference`.

certik commented on sympy/geometry/ellipse.py:
```diff
@@ -28,7 +28,7 @@
 from .line import Line, LinearEntity
 from .util import idiff
-
+from sympy.core.symbol import var
```
I think you should use relative imports, and not `var`, but `symbols`, see below.

certik reviewed (changes requested):
I left some comments.

certik commented:
Also, please try to write proper git commit messages. The first line has a column limit, then an empty line, then explaining what the commit does. After the PR is finished, let's try to write it as a set of small, logical commits.

certik commented:
Just search online for some guidelines. Here is one: https://chris.beams.io/posts/git-commit/, with explanations + motivations.

asmeurer commented:
We also have some stuff in our dev guide https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages.

avishrivastava11 commented:
@certik 
>Also, please try to write proper git commit messages. The first line has a column limit, then an empty line, then explaining what the commit does. After the PR is finished, let's try to write it as a set of small, logical commits.

I will keep that in mind Sir, for sure.

avishrivastava11 committed:
replaced 'var' with 'symbols', improved documentation

avishrivastava11 commented:
@certik  I have addressed your reviews in this commit (3fb3f47). Please take a look. Also, please let me know if the commit message of this commit (3fb3f47) was fine or do I need to improve more. Thanks

avishrivastava11 committed:
added pluralization for 3 geometry methods, fixing #7713

avishrivastava11 commented:
@debugger22 @akshayah3  @smichr  Can you take a look at this issue? The latest commit (7a75f18) solves issue #7713.

avishrivastava11 committed:
fixed travis errors of 7a75f18 commit

avishrivastava11 commented:
@certik  Can you take a look now?

avishrivastava11 commented:
@jksuom  Sir can you take a look?

Abdullahjavednesar commented on sympy/geometry/ellipse.py:
```diff
@@ -323,6 +322,39 @@ def circumference(self):
+        22.1034921607095
+        """
+
+        import math
```
You should rather import sqrt, pi, etc from sympy, no need to import math.

avishrivastava11 replied:
@Abdullahjavednesar   will do that.

avishrivastava11 committed:
used sympy functions instead of math - review by @Abdullahjavednesar

avishrivastava11 commented:
@Abdullahjavednesar  I've addressed your review. Please take a look. Also, since I've addressed what you requested, could you remove the `author's turn` label, as it's misleading?

jksuom commented:
Algebraic-geometric mean is also implemented in `mpmath` ([agm](https://github.com/fredrik-johansson/mpmath/blob/master/mpmath/function_docs.py#L4972-L4984)) and used to compute complete elliptic integrals. The use of native `mpf` type makes it more efficient than an implementation with `Float` objects in SymPy.
```
In [1]: e = Ellipse(Point(0,0), 4, 3)

In [2]: %time e.circumference.n()
CPU times: user 82.5 ms, sys: 16.3 ms, total: 98.8 ms
Wall time: 86.6 ms
Out[2]: 22.1034921607095

In [3]: %time e.ellipse_circumference()
CPU times: user 196 ms, sys: 12.4 ms, total: 208 ms
Wall time: 202 ms
Out[3]: 22.1034921607095
```
Is there any reason to include this in SymPy?

avishrivastava11 commented:
@jksuom  Sorry but I could not understand what you are trying to say here. I added these functions as they made the calculation faster (which you showed earlier :-
```
In [1]: e = Ellipse(Point(0,0), 4, 3)

In [2]: %time e.circumference.n()
CPU times: user 82.5 ms, sys: 16.3 ms, total: 98.8 ms
Wall time: 86.6 ms
Out[2]: 22.1034921607095

In [3]: %time e.ellipse_circumference()
CPU times: user 196 ms, sys: 12.4 ms, total: 208 ms
Wall time: 202 ms
Out[3]: 22.1034921607095
```
)
Could you please clarify what is it that you are trying to say here?

avishrivastava11 commented:
@jksuom  Sorry, but I had initially thought that `.ellipse_circumference()` was faster than `.circumference` method. If it's the other way around, I'll remove that method (from the entire code of this PR).  Should I do that ?
Also, what do you think about the other methods I added namely `def equation_using_slope(self, slope):` and `def are_collinear(*args):` ?

smichr committed:
make coefficient detection more robust

Although linsolve expands expressions via Poly and detects non-linearity
before passing to linear_eq_to_matrix, the latter does its own non-linearity
checking without using expansion so it will detect removable singularities
and systems that have not been simplified.

avishrivastava11 commented:
@jksuom ping

jksuom commented:
`solve` is an expensive function. I would use the precomputed coefficients (as solved in the SO [comment](https://math.stackexchange.com/questions/108270/what-is-the-equation-of-an-ellipse-hat-is-not-aligned-with-the-axis/646971)) instead of repeatedly calling `solve`.

avishrivastava11 committed:
removed usage of 'solve' from 'equation_using_slope' method

avishrivastava11 committed:
fixing travis errors by modifying docstring

avishrivastava11 commented:
@jksuom  I've done that now sir. It turns out there was no need for even a single `solve` statement.
>I had initially thought that .ellipse_circumference() was faster than .circumference method. If it's the other way around, I'll remove that method

Should I remove it?
Please take a look.

avishrivastava11 commented:
@certik Can you take a look?

avishrivastava11 committed:
modified IndexError for only integers-@sidhantnagpal 's review

sylee957 committed:
iszerofunc parameter for nullspace()

asmeurer committed:
Merge pull request #15162 from asmeurer/printing-imports

Change the printing imports in sympy/__init__.py to import from *

sylee957 committed:
iszerofunc parameter for eigenvects()

moorepants committed:
Added an example of an autolev/sympy output test.

moorepants committed:
Express not needed, done in parser output.

asmeurer committed:
Correctly handle defining a Singleton class with the same name twice

Without this, there are doctest failures from the Singleton doctests when its
run twice in the same Python session (doctest('singleton', subprocess=False)).

There are still some side effects from running the tests or doctests with
subprocess=False, in that the test singleton classes are registered on the S
object.

Fixes a failure from #15149.

valglad committed:
use max_cosets instead of changing CosetTable.coset_table_max_limit

Before introduction of the keyword `max_cosets`, the only way to
control the maximum number of cosets in a coset enumeration was
by changing the class attribute `coset_table_max_limit`. If not
restored to the original value, the change in attribute affects
all subsequent coset enumerations causing doctests to fail if run
twice in the same process. This commit changes the tests to use
`max_cosets` instead but also modifies the code of CosetTable to
allow the attribute change to be preserved even if `max_cosets`
is used after the change

moorepants committed:
Added angular velocity check [skip ci].

valglad committed:
use copies of imported Polyhedron objects in doctests

asmeurer committed:
Remove reference to NO ENTRY from the pull request template

It now mentioned only in the guide on the wiki. This is to make it less likely
for people to wrote "NO ENTRY" on pull requests that need a release notes
entry.

Fixes sympy/sympy-bot#30.

asmeurer committed:
Merge pull request #15167 from valglad/double_tests

fix combinatorics failures from repeated test runs

sylee957 committed:
Suggestions from @moorepants

cbm755 committed:
Merge pull request #15159 from sudz123/Issue_15097

Solvers: used checksysodesol for linear_3eq_order1 in test_ode.py

avishrivastava11 committed:
added changes from review by @sidhantnagpal

jksuom committed:
Merge pull request #15169 from sylee957/myCommit

Unmask iszerofunc parameter for nullspace() and eigenvects()

smichr committed:
Merge pull request #15168 from asmeurer/no-entry

Remove reference to NO ENTRY from the pull request template

NikhilPappu committed:
Made the requested changes

Also fixed the position vector error observed in #15165.

avishrivastava11 committed:
removed try: statement - @smichr review

avishrivastava11 committed:
removed try: statement - @smichr review

avishrivastava11 commented:
@Abdullahjavednesar  Can you take a look?

NikhilPappu committed:
Removed kane.py changes and replaced the commented code

jksuom committed:
Add clear_cache to CRootOf

Also call clear_cache to reset doctests in rootoftools and rootisolation.

NikhilPappu committed:
Cleaned up the doctest

NikhilPappu committed:
Added doctest_depends_on decorator

NikhilPappu committed:
Added doctest_depends_on decorator

asmeurer committed:
Move the super() call to the end of FunctionClass.__init__

The super() call calls BasicMeta.__init__, which adds the class to a set of
classes (sympy.core.core.all_classes). This computes the hash of the class.
However, the hash of a Function depends on its nargs (class_key()), meaning if
it is set dynamically, the hash will change and the class won't be found in
the all_classes set any more, breaking sympify(obj).

This happens for things like Function('f', eval=lambda self, t: ...) where
nargs is determined from the eval lambda.

Fixes #15170.

asmeurer committed:
Fix a test so it works in Python 2

moorepants committed:
Merge branch 'autolev_parser' of github.com:NikhilPappu/sympy into pr-15006

moorepants committed:
Corrected the expected vector to match the autolev syntax.

moorepants committed:
Merge pull request #15006 from NikhilPappu/autolev_parser

Updated parser code

Upabjojr committed:
renamed .integrate( ) to .compute_expectation( ) in sympy.stats

sylee957 committed:
First Draft

jksuom committed:
Merge pull request #15176 from Upabjojr/rename_stats_integrate

renamed .integrate( ) to .compute_expectation( ) in sympy.stats

asmeurer committed:
Fix arguments to super().__init__

The aren't actually used, but obviously it should be __init__(*args,
**kwargs), not __init__(args, kwargs).

asmeurer committed:
Unregister a test assumption handler at the end of the test

Otherwise it changes the behavior of Q.prime after the test has been run.

Related to issue #15149

asmeurer committed:
Don't reset the module counter in the autowrap tests

This messes up any further autowrap calls, as it gets confused with prior
modules. This fixes test failures from running the autowrap tests twice in the
same Python session (subprocess=False). See issue #15149.

jksuom committed:
Clear CRootOf cache for test_eval_approx_relative

sylee957 committed:
First paragraph

sylee957 committed:
Second Paragraph

sylee957 committed:
Doc revision

sylee957 committed:
Doc Revision

sylee957 committed:
Third Paragraph and examples

sylee957 committed:
Doc Revision

sylee957 committed:
Added Example

sylee957 committed:
Added paragraph

sylee957 committed:
Added Paragraph

smichr committed:
Merge pull request #15155 from avishrivastava11/row_column_insert_fix

Added implementation of IndexError for non int-indices in row_insert() and col_insert()

smichr commented on sympy/geometry/ellipse.py:
```diff
@@ -433,6 +465,45 @@ def equation(self, x='x', y='y'):
+        Requiring that the distance between the intersections of E and L be 2M identifies
+        b = M**2*(1 + s**2)
+
+        and similarly, requiring that the intersections between E and l be separated by 2m identifies
```
"E and l" -> "E and L"

smichr commented on sympy/geometry/ellipse.py:
```diff
@@ -433,6 +465,45 @@ def equation(self, x='x', y='y'):
+        and similarly, requiring that the intersections between E and l be separated by 2m identifies
+        a = m**2*(1 + s**2)
+
+        So the general equation of the ellipse centered at (xc,yc) whose major axis (with radius of M)
```
```
So the general equation of the ellipse centered at (xc,yc) whose major axis (with radius of M)
        is on a line with slope s, and whose minor axis has radius of m, is given by the solutions of:
         The ellipse is then defined by the zeros of
```

```
So the points that are on an ellipse centered at (xc,yc) whose major axis (with radius of M)
        is on a line with slope s, and whose minor axis has radius of m, are given be the zeros of:
```
It would be appropriate to reference this SO page. If you search SymPy for references to wkipedia you will find a suitable example.

avishrivastava11 committed:
shortened 'are_colliner' method, modified documentation-@smichr

avishrivastava11 committed:
shortened 'are_colliner' method, modified documentation-@smichr

jksuom committed:
Merge pull request #15174 from asmeurer/undefinedfunction-hash

Move the super() call to the end of FunctionClass.__init__

sylee957 committed:
Completed draft

sylee957 committed:
Doc Revision

sylee957 committed:
Doc Revision

sylee957 committed:
Added Footnote

sylee957 committed:
Added Footnote

sylee957 committed:
Footnote Revision

sylee957 committed:
Footnote Revision

sylee957 committed:
Added Footnote

sylee957 committed:
Added Footnote

sylee957 committed:
EOF with newline

rdbliss committed:
Sequences: Implement recursive sequence subclass

Introduces a new SeqBase subclass, `RecursiveSeq`, for sequences defined
recursively. Specifically, sequences that depend on a finite, fixed
number of their previous terms are supported. Any recurrence that is
expressible as a SymPy expression should work. This class will make it
easier to implement specialized sequences with more interesting
properties, such as linear recurrence relations.

Examples:

In [1]: from sympy.abc import n

In [2]: from sympy import Function

In [3]: y = Function("y")

In [4]: from sympy.series.sequences import RecursiveSeq

In [5]: fib = RecursiveSeq([0, 1], y(n), y(n - 1) + y(n - 2))

In [6]: fib[:10]
Out[6]: [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

In [7]: fib.degree
Out[7]: 2

In [8]: fib.recurrence
Out[8]: y(n - 2) + y(n - 1)

In [9]: from sympy import cos, sin

In [10]: more_complex = RecursiveSeq([0, 1], y(n), cos(y(n - 1)) + sin(y(n - 2)))

In [11]: more_complex[:5]
Out[11]: [0, 1, cos(1), sin(1) + cos(cos(1)), cos(sin(1) + cos(cos(1))) + sin(cos(1))]

sylee957 committed:
Modified output with pretty print

sylee957 committed:
Fixed typo and some other changes

sylee957 committed:
Added doctest skip to warning printings

sylee957 committed:
Relocated doctest skip tag

sylee957 committed:
Reformatted to 80-char max lines

sylee957 committed:
Trimmed indentation

rdbliss committed:
Add test for RecursiveSeq and make it pass

smichr committed:
fixes #15171: boolmap

rdbliss committed:
Change __init__ to __new__ in recursive sequence

This makes the args-test actually pass for RecursiveSeq.

avishrivastava11 commented:
@Abdullahjavednesar @jksuom @ArifAhmed1995 Can you restart this failing test 28715.22 of ea3f48c i.e. the last commit ? ( due to restarting 28715.21 passed, but 28715.22 still fails 
)

sylee957 committed:
Divided plot_and_save_1

jksuom committed:
Merge pull request #15181 from sylee957/docs

New tutorial section addressing iszerofunc from matrices.

avishrivastava11 commented:
does this look good to merge now?

sylee957 committed:
Divided test_matplotlib_1

sylee957 committed:
Divided plot_and_save_2

sylee957 committed:
Divided test_matplotlib_2

sylee957 committed:
Divided plot_and_save_3

sylee957 committed:
Divided test_matplotlib_3

sylee957 committed:
Divided plot_and_save_4

sylee957 committed:
Divided test_matplotlib_4

asmeurer committed:
Merge pull request #15139 from asmeurer/skip-cloudpickle

Skip cloudpickle test for pickling Function('f')

asmeurer committed:
Add a warning to the IPython tests about

Based on the discussion at #15149, there isn't a good way to make the IPython
tests not modify the existing IPython session, or to not affect itself when
run twice in the same Python session (test('ipython', subprocess=False)),
because IPython uses a single shell instance which gets modified by the test.

sylee957 committed:
Divided plot_and_save_5 and test_matplotlib_5

moorepants committed:
Merge branch 'master' into autolev-output-test

moorepants committed:
Updated note on definition of P_O_E>.

moorepants committed:
Merge pull request #14821 from cbm755/octave_codegen2

Octave/Matlab codegen improvements

rdbliss committed:
Return equation from RecursiveSeq for recurrence

This also adds some docstrings to clarify what the argument for
`recurrence` should be.

sylee957 committed:
Modified 5 to 6 hanoi

sylee957 committed:
Splitted test 4 because of found bottleneck

moorepants committed:
Moved new test to parsing.tests.

moorepants committed:
PEP8 fixes and other formatting.

smichr committed:
Update rootisolation.py

smichr committed:
Update rootoftools.py

asmeurer committed:
Merge pull request #15175 from jksuom/rootoftools-clear-cache

Add clear_cache to CRootOf

asmeurer committed:
Merge branch 'master' into repeated-test-failures

asmeurer committed:
Merge pull request #15163 from asmeurer/repeated-test-failures

Fix test failures from repeated test runs

jksuom committed:
Merge pull request #15191 from sylee957/partition-plot-test

Partitioned test_matplotlib into chunks

moorepants committed:
Merge pull request #15165 from moorepants/autolev-output-test

Example of an Autolev output test

asmeurer committed:
Bump version to 1.3rc1

asmeurer committed:
Bump version to 1.4.dev

asmeurer committed:
Update AUTHORS and .mailmap

asmeurer committed:
Merge pull request #15197 from asmeurer/version-bump

Bump version to 1.4.dev

asmeurer committed:
Make sure all printing settings get reset between tests/doctests

rdbliss committed:
Clean RecursiveSeq docstring

rdbliss committed:
Improve logic in RecursiveSeq creation

rdbliss committed:
Reorder arguments in RecursiveSeq

This also implements default dummy initial values.

rdbliss committed:
Fix index calculations in RecursiveSeq

sylee957 committed:
Delete unused import

rdbliss committed:
Remove stop argument to RecursiveSeq

It is now up to the user to determine any undefined or infinite return
values from their recurrence.

rdbliss committed:
Change RecursiveSeq.gen to RecursiveSeq.__iter__

asmeurer committed:
Set Permutation.print_cyclic = True in the homomorphisms doctest

This works around an issue where running the permutations.py doctests first
changes this flag and causes the homomorphisms doctests to fail.

This is just a workaround for a larger issue, which is described in #15198 and

Abdullahjavednesar committed:
Merge pull request #15199 from sylee957/codeclean

Delete unused import

asmeurer committed:
Remove rubi_parsing_guide.md

It has been moved to the wiki, which is a better place for developer
documentation.

asmeurer committed:
Update release tarball git whitelist

rdbliss committed:
Make function argument explicit in RecursiveSeq

sylee957 committed:
Unmask iszerofunc in det()

sylee957 committed:
Unmask iszerofunc in _eval_det_bareiss()

sylee957 committed:
Revert "Unmask iszerofunc in det()"

This reverts commit b90f9ed6efc1bcb625dd03f518d8b69fb3d9ff86.

rdbliss committed:
Fix RecurisveSeq args to be instances of Basic

rdbliss committed:
RecursiveSeq: Use S.Infinity over oo

sylee957 committed:
Conflict in det()

sylee957 committed:
iszerofunc for det() mapped with None

sylee957 committed:
Keyword validity check

rdbliss committed:
RecursiveSeq: Update docstring

smichr committed:
linear_coeffs added to solveset

In addition, nonlinearity in Eq which would
cancel by subtraction is now detected in
linear_eq_to_matrix.

Note: the requirement that symbols be Symbols
is made to avoid problems like finding the
coefficients of x in `x + f(x)`. `as_coeff_mul(x)`
will treat both terms as x-dependent and will not
return f(x) as a constant and 1 as the coefficient.

jksuom committed:
Merge pull request #15184 from rwbogl/recursive-sequences

Implement expression-based recursive sequence

smichr committed:
make sure Eq simplification happens in linsolve

asmeurer committed:
Merge pull request #14949 from jlumpe/theano-printer-fixes

Clean up Theano printer, add comments and docstrings

jksuom committed:
Merge pull request #15207 from sylee957/unmask_iszerofunc

Unmask iszerofunc from _eval_det_bareiss() and det()

cbm755 committed:
codegen: revert to printing comments with unknown fcns

Recent commits added a new `allow_unknown_functions` flag to
CodePrinter and children.  Change the default of this flag to False
to revert to the previous behaviour.  Update tests and documentation.

`lambdify` apparently needs this new flag for #15071 so turn on the
flag there.

cbm755 committed:
codegen: add test ensuring Octave's whitelist is respected

A recent change for meant things not on the whitelist were nonetheless
printed.  Add a test for a function that is not on our whitelist to
ensure such a change would be caught.

Fixes #15206.  Or more precisely make sure its not broken again.

smichr committed:
review edits

smichr committed:
added test for issue #15195

asmeurer committed:
Merge pull request #15215 from cbm755/codegen_unknown_fcn_fix

codegen: revert to printing comments with unknown fcns

sylee957 committed:
Merged determinant test

sylee957 committed:
Added missing test

smichr committed:
Application & Subs mods

* Application recurses into args while doing subs
* Subs.has gives structural (not mathematical) answer
* Subs(Subs will denest
* Subs includes null subs as part of hash and __eq__ testing

cbm755 committed:
Merge pull request #15212 from smichr/patch-19

linear_coeffs function added to solveset

smichr committed:
Sum gets dummy_eq method

smichr committed:
_atomic gets recursive flag

smichr committed:
remove unused multinomial0

multinomial_coefficients0 is similar to the
version without the appended 0 but is unused in
SymPy and gives negative k values, e.g.

>>> (list(multinomial_coefficients0(3,1)))
[(1, 0, 0), (0, 1, 0), (0, 0, 1), (-1, 1, 1)]

smichr committed:
Poly.replace note and handling of non-gen replacement

smichr committed:
ordinal(1)->1st: ordinal added to misc

maurogaravello committed:
Moved the import statement at the beginning of heuristics() function

moorepants committed:
Added the Mod() function to the Fortran printer, fixes 15230.

moorepants committed:
Added test for the autowrap command reported in the issue.

moorepants committed:
skip test if f2py isn't installed.

Upabjojr committed:
refactored index contraction algorithms for TensMul

raineszm committed:
exprtools: modify _mask_nc for functions

As pointed out in issue 8263, _mask_nc treats noncommutative objects
with all commutative arguments as commutative. This commit changes
_mask_nc to replace any object for which is_commutative is False with a
noncommutative Dummy.

asmeurer committed:
Unpin the sage version in Travis

asmeurer committed:
Unpin the sage version in Travis

asmeurer committed:
Install sagelib instead of the full sage on Travis

This should hopefully be sufficient for the tests, and fix the "no space left
on device" error from installing the full sage.

asmeurer committed:
Install sagelib instead of the full sage on Travis

This should hopefully be sufficient for the tests, and fix the "no space left
on device" error from installing the full sage.

isuruf committed:
Comment removing sympy

sagelib doesn't pull in sympy, so no need to remove it

isuruf committed:
Comment removing sympy

sagelib doesn't pull in sympy, so no need to remove it

asmeurer committed:
Don't need to remove sympy when we are just installing sagelib

asmeurer committed:
Don't need to remove sympy when we are just installing sagelib

asmeurer committed:
Install mpmath in the Sage environment

asmeurer committed:
Install mpmath in the Sage environment

asmeurer committed:
Merge branch 'sage-fix' of github.com:asmeurer/sympy into sage-fix

asmeurer committed:
Fix IPython pinning in the Sage environment on Travis

asmeurer committed:
Fix IPython pinning in the Sage environment on Travis

moorepants committed:
Switch to modulo and check sign.

moorepants committed:
mod -> modulo in tests.

asmeurer committed:
Sage installs in Python 3 now

asmeurer committed:
Sage installs in Python 3 now

jksuom committed:
Merge pull request #15233 from asmeurer/sage-fix

Fix Travis "no space left on devices" Sage build issues

sylee957 committed:
Merge branch 'new' into merge_matrix_test
revert commit

jksuom committed:
Merge pull request #15221 from sylee957/merge_matrix_test

Merge some matrix tests

smichr committed:
Update function.py

moorepants committed:
Merge branch 'master' into issue-15230-fortran-mod

sylee957 committed:
Added test case for #15207

asmeurer committed:
Change SymPy version to 1.3 (final)

asmeurer committed:
Merge branch 'master' into 1.3

smichr committed:
Merge pull request #15225 from smichr/boolmap

boolmap doesn't confuse Xor and ~Xor

moorepants committed:
Fortran Mod() printing raises error if F66 or F77.

- Added _print_Mod() that raises error if invalid Fortran versions.
- Added tests for all standards for Mod() printing.
- Forced autowrap to use the F95 standard to match the statements in the
  docstrings.

moorepants committed:
Added test for integer arguments, removed tempdir.

sylee957 committed:
Changed test cases to 4*4 matrix

avishrivastava11 commented:
@certik @smichr does this look good to merge now?

moorepants committed:
ValueError -> NotImplementedError and remove evalf() to retain integer.

sylee957 committed:
Added tutorial section for Win 10 users

sylee957 committed:
Fixed some unsupported markdown

sylee957 committed:
Code into hyperlink markdown

smichr committed:
Merge pull request #15235 from sylee957/add_test

Added test case for iszerofunc in det

moorepants committed:
Changed to NotImplementedError in the tests.

sylee957 committed:
Revised tutorial

Upabjojr committed:
TensAdd is no longer flattened in TensMul; canonicalization on object construction disabled for TensAdd

smichr committed:
Merge pull request #15222 from smichr/d1

Application & Subs mods

smichr committed:
Merge pull request #14927 from schymans/si-extension

Add derived dimensions to si.py.

sylee957 committed:
Fixed typo and backslash

Upabjojr committed:
added pretty printer for tensors

asmeurer committed:
Fix missing newline in release.py

cbm755 committed:
Merge pull request #15198 from sympy/1.3

1.3 release branch

Upabjojr committed:
pprint for TensorIndex

Upabjojr committed:
fixed gamma matrices tests

Upabjojr committed:
latex printers for tensors

Abdullahjavednesar committed:
Merge pull request #15248 from sylee957/add_sphinx_tutorial_windows

Add doc build tutorial for windows users.

Upabjojr committed:
Merge pull request #15250 from Upabjojr/tensor_subclass_expr

Remove automatic expansions of tensorial expressions.

Upabjojr committed:
Merge pull request #15249 from Upabjojr/tensor_pretty_print

added pretty printer for tensors

smichr committed:
Merge pull request #15231 from moorepants/issue-15230-fortran-mod

Support Mod with fcode

Upabjojr committed:
fixing LaTeX printer; fixes to str and pretty printers as well

Upabjojr committed:
fixing bug in LaTeX printer for tensors

Upabjojr committed:
Tensor now requires .doit() to contract indices

Upabjojr committed:
Added .doit() for TensMul

valglad committed:
Merge pull request #14939 from maurogaravello/14574_limit

limit(sqrt(x)*cos(x - x**2) / (x + 1), x, oo) causes recursion error

Upabjojr committed:
Added .doit() for TensAdd

Upabjojr committed:
Merge pull request #15251 from Upabjojr/tensor_latex_print

latex printers for tensors

Upabjojr committed:
merge master

Upabjojr committed:
fixing printers for indices

Upabjojr committed:
Merge branch 'fix_tensor_pprint_index' into tensor_mixed_mul_add_subexpr

Upabjojr committed:
Merge pull request #15253 from Upabjojr/fix_tensor_pprint_index

fixing printers for indices

Upabjojr committed:
Option for optional index replacements; fixing .expand() for tensors.

Upabjojr committed:
added .expand() tests to tensors; fixed gamma matrix tests

Upabjojr committed:
Merge pull request #15252 from Upabjojr/tensor_mixed_mul_add_subexpr

Tensor: support mixed subexpressions of additions and multiplications

smichr committed:
Merge pull request #15099 from asmeurer/sympify-numpy-args

Pass sympify args through when sympifying numpy types

smichr commented on sympy/geometry/ellipse.py:
```diff
@@ -433,6 +465,49 @@ def equation(self, x='x', y='y'):
+        a = m**2*(1 + s**2)
+
+        So the points that are on an ellipse centered at (xc,yc) whose major axis (with radius of M)
+        is on a line with slope s, and whose minor axis has radius of m, are given be the zeros of:
```
be->by

avishrivastava11 replied:
@smichr Done now. Please take a look.

smichr commented on sympy/geometry/ellipse.py:
```diff
@@ -433,6 +465,49 @@ def equation(self, x='x', y='y'):
+        Requiring that the distance between the intersections of E and L be 2M identifies
+        b = M**2*(1 + s**2)
+
+        and similarly, requiring that the intersections between E and L be separated by 2m identifies
```
I would suggest keeping the structure parallel

-> and similarly, requiring that the distance between the intersections of E and l be 2m identifies

A minimum change needed is to change L to l.

avishrivastava11 replied:
@smichr Done now. Please take a look.

grozin committed:
functions/special/zeta_functions.py: fix docstring in eta

avishrivastava11 committed:
minor changes in documentation - @smichr review

avishrivastava11 committed:
minor changes in documentation - @smichr review

smichr committed:
Merge pull request #15246 from smichr/d9

Poly.replace note and handling of non-gen replacement

smichr committed:
Merge pull request #15242 from smichr/d4

_atomic gets recurse flag

jksuom committed:
Merge pull request #15260 from grozin/master

functions/special/zeta_functions.py: fix docstring in eta

Upabjojr committed:
Shortcut for tensor indices

Upabjojr committed:
Merge pull request #15262 from Upabjojr/tensor_easier_indices

Shortcut for tensor indices

avishrivastava11 committed:
added extra parameter 'slope' to 'equation' method-@smichr review

avishrivastava11 committed:
added extra parameter 'slope' to 'equation' method-@smichr review

gdp1807 committed:
Merge git://github.com/sympy/sympy

gdp1807 committed:
modified Circle to return Point when radius is zero

smichr commented on sympy/geometry/ellipse.py:
```diff
@@ -418,20 +475,47 @@ def equation(self, x='x', y='y'):
+        (-x + y)**2/32 + (x + y)**2/2 - 1
 
         """
-        x = _symbol(x, real=True)
```
x and y should stay here and not be below in the else

avishrivastava11 replied:
@smichr Done. Please take a look.

smichr commented on sympy/geometry/ellipse.py:
```diff
@@ -418,20 +475,47 @@ def equation(self, x='x', y='y'):
-        return t1 + t2 - 1
+
+        if slope is not None:
+            a, b, x, y, x_c, y_c = symbols('a,b,x,y,x_c,y_c')
```
none of these are needed

avishrivastava11 replied:
@smichr Done. Please take a look.

smichr commented on sympy/geometry/ellipse.py:
```diff
@@ -418,20 +475,47 @@ def equation(self, x='x', y='y'):
+            L = ((y - y_c) - slope * (x - x_c)) ** 2
+            l = (slope * (y - y_c) + (x - x_c)) ** 2
+
+            if idiff(L, y, x) == -1 / idiff(l, y, x):
```
no need for if, just return the result AFAICT

avishrivastava11 replied:
@smichr Done. Please take a look.

sylee957 committed:
Sanitize plot xlim

sylee957 committed:
Sanitize plot ylim

smichr committed:
Merge pull request #15245 from smichr/d8

remove unused multinomial0

czgdp1807 committed:
Added test_issue_15259

smichr committed:
Merge pull request #15265 from sylee957/plot_fix

Add float sanitization to xlim and ylim in plot

smichr committed:
Merge pull request #15264 from czgdp1807/patch-32

Modified Circle to return Point when radius is zero

avishrivastava11 committed:
shortened code and modified tests by comparing 'str'

avishrivastava11 committed:
shortened code and modified tests by comparing 'str'

avishrivastava11 committed:
fixing travis errors

avishrivastava11 committed:
fixing travis errors

asmeurer committed:
Merge pull request #15247 from smichr/d10

ordinal(1)->1st: ordinal added to misc

asmeurer committed:
Merge pull request #15239 from smichr/d2

Sum gets dummy_eq method

jksuom committed:
Merge pull request #15232 from raineszm/issue-8263

exprtools: modify _mask_nc for functions

avishrivastava11 committed:
Merge branch 'avishrivastava11_ellipse_eq' of https://github.com/avishrivastava11/sympy into avishrivastava11_ellipse_eq

avishrivastava11 committed:
fixing travis errors and removing unnecessary brackets

avishrivastava11 commented:
@smichr  Looks like the commits got messed up somehow. I'll be opening up a new PR for this implementation (sort of a continuation) once [this](https://github.com/sympy/sympy/pull/15273) PR gets merged. Will that be fine?

smichr commented:
continued in #15286
